File: perl/tests/searcher_test.pl
#!/usr/bin/perl -w
#
# searcher_test.pl
#
#
use strict;
use warnings;

use Test::Simple tests => 73;

use lib '/Users/cary/src/git/xsearch/perl';

use plsearch::SearchSettings;
use plsearch::Searcher;


sub get_settings {
    my $settings = new plsearch::SearchSettings();
    $settings->{startpath} = '.';
    push(@{$settings->{searchpatterns}}, "Searcher");
    return $settings;
}

sub test_validate_settings {
    my $settings = get_settings();
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
}


################################################################################
# is_search_dir tests
################################################################################

sub test_is_search_dir_no_patterns {
    my $settings = get_settings();
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $dir = 'plsearch';
    ok($searcher->is_search_dir($dir), "$dir is search dir with no patterns");
}

sub test_is_search_dir_matches_in_pattern {
    my $settings = get_settings();
    push(@{$settings->{in_dirpatterns}}, 'plsearch');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $dir = 'plsearch';
    ok($searcher->is_search_dir($dir), "$dir matches in_dirpatterns");
}

sub test_is_search_dir_no_match_in_pattern {
    my $settings = get_settings();
    push(@{$settings->{in_dirpatterns}}, 'plsearch');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $dir = 'pysearch';
    ok(!$searcher->is_search_dir($dir), "$dir does not match in_dirpatterns");
}

sub test_is_search_dir_matches_out_pattern {
    my $settings = get_settings();
    push(@{$settings->{out_dirpatterns}}, 'pysearch');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $dir = 'pysearch';
    ok(!$searcher->is_search_dir($dir), "$dir matches out_dirpatterns");
}

sub test_is_search_dir_no_match_out_pattern {
    my $settings = get_settings();
    push(@{$settings->{out_dirpatterns}}, 'pysearch');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $dir = 'plsearch';
    ok($searcher->is_search_dir($dir), "$dir does not match out_dirpatterns");
}

sub test_is_search_dir_single_dot {
    my $settings = get_settings();
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $dir = '.';
    ok($searcher->is_search_dir($dir), "$dir is search dir");
}

sub test_is_search_dir_double_dot {
    my $settings = get_settings();
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $dir = '..';
    ok($searcher->is_search_dir($dir), "$dir is search dir");
}

sub test_is_search_dir_hidden_dir {
    my $settings = get_settings();
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $dir = '.git';
    ok(!$searcher->is_search_dir($dir), "Hidden dir $dir is not search dir by default");
}

sub test_is_search_dir_hidden_dir_include_hidden {
    my $settings = get_settings();
    $settings->{excludehidden} = 0;
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $dir = '.git';
    ok($searcher->is_search_dir($dir),
        "Hidden dir $dir is search dir with excludehidden set to false");
}

################################################################################
# is_search_file tests
################################################################################

sub test_is_search_file_matches_by_default {
    my $settings = get_settings();
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'FileUtil.pm';
    ok($searcher->is_search_file($file), "$file is search file by default");
}

sub test_is_search_file_matches_in_extension {
    my $settings = get_settings();
    push(@{$settings->{in_extensions}}, 'pm');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'FileUtil.pm';
    ok($searcher->is_search_file($file), "$file matches in_extensions");
}

sub test_is_search_file_no_match_in_extension {
    my $settings = get_settings();
    push(@{$settings->{in_extensions}}, 'pl');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'FileUtil.pm';
    ok(!$searcher->is_search_file($file), "$file does not match in_extensions");
}

sub test_is_search_file_matches_out_extension {
    my $settings = get_settings();
    push(@{$settings->{out_extensions}}, 'pm');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'FileUtil.pm';
    ok(!$searcher->is_search_file($file), "$file matches out_extensions");
}

sub test_is_search_file_no_match_out_extension {
    my $settings = get_settings();
    push(@{$settings->{out_extensions}}, 'py');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'FileUtil.pm';
    ok($searcher->is_search_file($file), "$file does not match out_extensions");
}

sub test_is_search_file_matches_in_pattern {
    my $settings = get_settings();
    push(@{$settings->{in_filepatterns}}, 'Search');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'Searcher.pm';
    ok($searcher->is_search_file($file), "$file matches in_filepatterns");
}

sub test_is_search_file_no_match_in_pattern {
    my $settings = get_settings();
    push(@{$settings->{in_filepatterns}}, 'Search');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'FileUtil.pm';
    ok(!$searcher->is_search_file($file), "$file does not match in_filepatterns");
}

sub test_is_search_file_matches_out_pattern {
    my $settings = get_settings();
    push(@{$settings->{out_filepatterns}}, 'Search');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'Searcher.pm';
    ok(!$searcher->is_search_file($file), "$file matches out_filepatterns");
}

sub test_is_search_file_no_match_out_pattern {
    my $settings = get_settings();
    push(@{$settings->{out_filepatterns}}, 'Search');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'FileUtil.pm';
    ok($searcher->is_search_file($file), "$file does not match out_filepatterns");
}


################################################################################
# is__archive_search_file tests
################################################################################

sub test_is_archive_search_file_matches_by_default {
    my $settings = get_settings();
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    ok($searcher->is_archive_search_file($file), "$file is archive search file by default");
}

sub test_is_archive_search_file_matches_in_extension {
    my $settings = get_settings();
    push(@{$settings->{in_archiveextensions}}, 'zip');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    ok($searcher->is_archive_search_file($file), "$file matches in_archiveextensions");
}

sub test_is_archive_search_file_no_match_in_extension {
    my $settings = get_settings();
    push(@{$settings->{in_archiveextensions}}, 'gz');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    ok(!$searcher->is_archive_search_file($file), "$file does not match in_archiveextensions");
}

sub test_is_archive_search_file_matches_out_extension {
    my $settings = get_settings();
    push(@{$settings->{out_archiveextensions}}, 'zip');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    ok(!$searcher->is_archive_search_file($file), "$file matches out_archiveextensions");
}

sub test_is_archive_search_file_no_match_out_extension {
    my $settings = get_settings();
    push(@{$settings->{out_archiveextensions}}, 'gz');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    ok($searcher->is_archive_search_file($file), "$file does not match out_archiveextensions");
}

sub test_is_archive_search_file_matches_in_pattern {
    my $settings = get_settings();
    push(@{$settings->{in_archivefilepatterns}}, 'arch');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    ok($searcher->is_archive_search_file($file), "$file matches in_archivefilepatterns");
}

sub test_is_archive_search_file_no_match_in_pattern {
    my $settings = get_settings();
    push(@{$settings->{in_archivefilepatterns}}, 'archives');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    ok(!$searcher->is_archive_search_file($file), "$file does not match in_archivefilepatterns");
}

sub test_is_archive_search_file_matches_out_pattern {
    my $settings = get_settings();
    push(@{$settings->{out_archivefilepatterns}}, 'arch');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    ok(!$searcher->is_archive_search_file($file), "$file matches out_archivefilepatterns");
}

sub test_is_archive_search_file_no_match_out_pattern {
    my $settings = get_settings();
    push(@{$settings->{out_archivefilepatterns}}, 'archives');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    ok($searcher->is_archive_search_file($file), "$file does not match out_archivefilepatterns");
}


################################################################################
# filter_file tests
################################################################################

sub test_filter_file_matches_by_default {
    my $settings = get_settings();
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'FileUtil.pm';
    ok($searcher->filter_file($file), "$file passes filter_file by default");
}

sub test_filter_file_is_search_file {
    my $settings = get_settings();
    push(@{$settings->{in_extensions}}, 'pm');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'FileUtil.pm';
    ok($searcher->filter_file($file), "$file passes filter_file when is_search_file");
}

sub test_filter_file_not_is_search_file {
    my $settings = get_settings();
    push(@{$settings->{in_extensions}}, 'pl');
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'FileUtil.pm';
    ok(!$searcher->filter_file($file), "$file does not pass filter_file when !is_search_file");
}

sub test_filter_file_is_hidden_file {
    my $settings = get_settings();
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = '.gitignore';
    ok(!$searcher->filter_file($file), "$file does not pass filter_file when hidden");
}

sub test_filter_file_hidden_includehidden {
    my $settings = get_settings();
    $settings->{excludehidden} = 0;
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = '.gitignore';
    ok($searcher->filter_file($file), "$file passes filter_file when hidden and excludehidden=0");
}

sub test_filter_file_archive_no_searcharchives {
    my $settings = get_settings();
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    #print "searcher->is_archive_search_file(archive.zip): " . $searcher->is_archive_search_file('archive.zip') . "\n";
    ok(!$searcher->filter_file($file), "$file does not pass filter_file when searcharchives=0");
}

sub test_filter_file_archive_searcharchives {
    my $settings = get_settings();
    $settings->{searcharchives} = 1;
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    #print "searcher->is_archive_search_file(archive.zip): " . $searcher->is_archive_search_file('archive.zip') . "\n";
    ok($searcher->filter_file($file), "$file passes filter_file when searcharchives=1");
}

sub test_filter_file_archive_archivesonly {
    my $settings = get_settings();
    $settings->{archivesonly} = 1;
    $settings->{searcharchives} = 1;
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'archive.zip';
    #print "searcher->is_archive_search_file(archive.zip): " . $searcher->is_archive_search_file('archive.zip') . "\n";
    ok($searcher->filter_file($file), "$file passes filter_file when archivesonly=1");
}

sub test_filter_file_nonarchive_archivesonly {
    my $settings = get_settings();
    $settings->{archivesonly} = 1;
    $settings->{searcharchives} = 1;
    my ($searcher, $errs) = new plsearch::Searcher($settings);
    ok(scalar @{$errs} == 0, 'No errors from valid settings');
    my $file = 'FileUtil.pm';
    #print "searcher->is_archive_search_file(archive.zip): " . $searcher->is_archive_search_file('archive.zip') . "\n";
    ok(!$searcher->filter_file($file), "$file does not pass filter_file when archivesonly=1");
}


################################################################################
# main
################################################################################
sub main {

    test_validate_settings();                             # 1 test

    # is_search_dir tests
    test_is_search_dir_no_patterns();                     # 2 tests
    test_is_search_dir_matches_in_pattern();              # 2 tests
    test_is_search_dir_no_match_in_pattern();             # 2 tests
    test_is_search_dir_matches_out_pattern();             # 2 tests
    test_is_search_dir_no_match_out_pattern();            # 2 tests
    test_is_search_dir_single_dot();                      # 2 tests
    test_is_search_dir_double_dot();                      # 2 tests
    test_is_search_dir_hidden_dir();                      # 2 tests
    test_is_search_dir_hidden_dir_include_hidden();       # 2 tests

    # is_search_file tests
    test_is_search_file_matches_by_default();             # 2 tests
    test_is_search_file_matches_in_extension();           # 2 tests
    test_is_search_file_no_match_in_extension();          # 2 tests
    test_is_search_file_matches_out_extension();          # 2 tests
    test_is_search_file_no_match_out_extension();         # 2 tests
    test_is_search_file_matches_in_pattern();             # 2 tests
    test_is_search_file_no_match_in_pattern();            # 2 tests
    test_is_search_file_matches_out_pattern();            # 2 tests
    test_is_search_file_no_match_out_pattern();           # 2 tests

    # is_archive_search_file tests
    test_is_archive_search_file_matches_by_default();     # 2 tests
    test_is_archive_search_file_matches_in_extension();   # 2 tests
    test_is_archive_search_file_no_match_in_extension();  # 2 tests
    test_is_archive_search_file_matches_out_extension();  # 2 tests
    test_is_archive_search_file_no_match_out_extension(); # 2 tests
    test_is_archive_search_file_matches_in_pattern();     # 2 tests
    test_is_archive_search_file_no_match_in_pattern();    # 2 tests
    test_is_archive_search_file_matches_out_pattern();    # 2 tests
    test_is_archive_search_file_no_match_out_pattern();   # 2 tests

    # filter_file tests
    test_filter_file_matches_by_default();                # 2 tests
    test_filter_file_is_search_file();                    # 2 tests
    test_filter_file_not_is_search_file();                # 2 tests
    test_filter_file_is_hidden_file();                    # 2 tests
    test_filter_file_hidden_includehidden();              # 2 tests
    test_filter_file_archive_no_searcharchives();         # 2 tests
    test_filter_file_archive_searcharchives();            # 2 tests
    test_filter_file_archive_archivesonly();              # 2 tests
    test_filter_file_nonarchive_archivesonly();           # 2 tests

}

main();

File: ruby/tests/searcher_test.rb
################################################################################
#
# searcher_test.rb
#
# Searcher testing
#
################################################################################

require_relative "../rbsearch/config.rb"
require_relative "../rbsearch/filetypes.rb"
require_relative "../rbsearch/searcher.rb"
require_relative "../rbsearch/searchsettings.rb"
require "test/unit"
 
class SearcherTest < Test::Unit::TestCase

  def get_settings
    settings = SearchSettings.new()
    settings.startpath = '.'
    settings.add_pattern('Searcher', settings.searchpatterns)
    settings
  end

  def get_test_file
    File.expand_path("#{SHAREDPATH}/testFiles/testFile2.txt")
  end

################################################################################
# is_search_dir tests
################################################################################
  def test_is_search_dir_no_patterns
    settings = self.get_settings()
    searcher = Searcher.new(settings)
    dir = 'plsearch'
    assert(searcher.is_search_dir(dir))
  end

  def test_is_search_dir_matches_in_pattern
    settings = self.get_settings()
    settings.add_pattern('plsearch', settings.in_dirpatterns)
    searcher = Searcher.new(settings)
    dir = 'plsearch'
    assert(searcher.is_search_dir(dir))
  end

  def test_is_search_dir_no_match_in_pattern
    settings = self.get_settings()
    settings.add_pattern('plsearch', settings.in_dirpatterns)
    searcher = Searcher.new(settings)
    dir = 'pysearch'
    assert(!searcher.is_search_dir(dir))
  end

  def test_is_search_dir_matches_out_pattern
    settings = self.get_settings()
    settings.add_pattern('pysearch', settings.out_dirpatterns)
    searcher = Searcher.new(settings)
    dir = 'pysearch'
    assert(!searcher.is_search_dir(dir))
  end

  def test_is_search_dir_no_match_out_pattern
    settings = self.get_settings()
    settings.add_pattern('pysearch', settings.out_dirpatterns)
    searcher = Searcher.new(settings)
    dir = 'plsearch'
    assert(searcher.is_search_dir(dir))
  end

  def test_is_search_dir_single_dot
    settings = self.get_settings()
    searcher = Searcher.new(settings)
    dir = '.'
    assert(searcher.is_search_dir(dir))
  end

  def test_is_search_dir_double_dot
    settings = self.get_settings()
    searcher = Searcher.new(settings)
    dir = '..'
    assert(searcher.is_search_dir(dir))
  end

  def test_is_search_dir_hidden_dir
    settings = self.get_settings()
    searcher = Searcher.new(settings)
    dir = '.git'
    assert(!searcher.is_search_dir(dir))
  end

  def test_is_search_dir_hidden_dir_include_hidden
    settings = self.get_settings()
    settings.excludehidden = false
    searcher = Searcher.new(settings)
    dir = '.git'
    assert(searcher.is_search_dir(dir))
  end

################################################################################
# is_search_file tests
################################################################################
  def test_is_search_file_matches_by_default
    settings = self.get_settings()
    searcher = Searcher.new(settings)
    f = 'fileutil.rb'
    assert(searcher.is_search_file(f))
  end

  def test_is_search_file_matches_in_extension
    settings = self.get_settings()
    settings.add_exts('rb', settings.in_extensions)
    searcher = Searcher.new(settings)
    f = 'fileutil.rb'
    assert(searcher.is_search_file(f))
  end

  def test_is_search_file_no_match_in_extension
    settings = self.get_settings()
    settings.add_exts('py', settings.in_extensions)
    searcher = Searcher.new(settings)
    f = 'fileutil.rb'
    assert(!searcher.is_search_file(f))
  end

  def test_is_search_file_matches_out_extension
    settings = self.get_settings()
    settings.add_exts('rb', settings.out_extensions)
    searcher = Searcher.new(settings)
    f = 'fileutil.rb'
    assert(!searcher.is_search_file(f))
  end

  def test_is_search_file_no_match_out_extension
    settings = self.get_settings()
    settings.add_exts('py', settings.out_extensions)
    searcher = Searcher.new(settings)
    f = 'fileutil.rb'
    assert(searcher.is_search_file(f))
  end

  def test_is_search_file_matches_in_pattern
    settings = self.get_settings()
    settings.add_pattern('search', settings.in_filepatterns)
    searcher = Searcher.new(settings)
    f = 'searcher.rb'
    assert(searcher.is_search_file(f))
  end

  def test_is_search_file_no_match_in_pattern
    settings = self.get_settings()
    settings.add_pattern('search', settings.in_filepatterns)
    searcher = Searcher.new(settings)
    f = 'fileutil.rb'
    assert(!searcher.is_search_file(f))
  end

  def test_is_search_file_matches_out_pattern
    settings = self.get_settings()
    settings.add_pattern('search', settings.out_filepatterns)
    searcher = Searcher.new(settings)
    f = 'searcher.rb'
    assert(!searcher.is_search_file(f))
  end

  def test_is_search_file_no_match_out_pattern
    settings = self.get_settings()
    settings.add_pattern('search', settings.out_filepatterns)
    searcher = Searcher.new(settings)
    f = 'fileutil.rb'
    assert(searcher.is_search_file(f))
  end

################################################################################
# is__archive_search_file tests
################################################################################
  def test_is_archive_search_file_matches_by_default
    settings = self.get_settings()
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(searcher.is_archive_search_file(f))
  end

  def test_is_archive_search_file_matches_in_extension
    settings = self.get_settings()
    settings.add_exts('zip', settings.in_archiveextensions)
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(searcher.is_archive_search_file(f))
  end

  def test_is_archive_search_file_no_match_in_extension
    settings = self.get_settings()
    settings.add_exts('gz', settings.in_archiveextensions)
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(!searcher.is_archive_search_file(f))
  end

  def test_is_archive_search_file_matches_out_extension
    settings = self.get_settings()
    settings.add_exts('zip', settings.out_archiveextensions)
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(!searcher.is_archive_search_file(f))
  end

  def test_is_archive_search_file_no_match_out_extension
    settings = self.get_settings()
    settings.add_exts('gz', settings.out_archiveextensions)
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(searcher.is_archive_search_file(f))
  end

  def test_is_archive_search_file_matches_in_pattern
    settings = self.get_settings()
    settings.add_pattern('arch', settings.in_archivefilepatterns)
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(searcher.is_archive_search_file(f))
  end

  def test_is_archive_search_file_no_match_in_pattern
    settings = self.get_settings()
    settings.add_pattern('archives', settings.in_archivefilepatterns)
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(!searcher.is_archive_search_file(f))
  end

  def test_is_archive_search_file_matches_out_pattern
    settings = self.get_settings()
    settings.add_pattern('arch', settings.out_archivefilepatterns)
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(!searcher.is_archive_search_file(f))
  end

  def test_is_archive_search_file_no_match_out_pattern
    settings = self.get_settings()
    settings.add_pattern('archives', settings.out_archivefilepatterns)
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(searcher.is_archive_search_file(f))
  end

################################################################################
# filter_file tests
################################################################################
  def test_filter_file_matches_by_default
    settings = self.get_settings()
    searcher = Searcher.new(settings)
    f = 'fileutil.rb'
    assert(searcher.filter_file(f))
  end

  def test_filter_file_is_search_file
    settings = self.get_settings()
    settings.add_exts('rb', settings.in_extensions)
    searcher = Searcher.new(settings)
    f = 'fileutil.rb'
    assert(searcher.filter_file(f))
  end

  def test_filter_file_not_is_search_file
    settings = self.get_settings()
    settings.add_exts('pl', settings.in_extensions)
    searcher = Searcher.new(settings)
    f = 'fileutil.rb'
    assert(!searcher.filter_file(f))
  end

  def test_filter_file_is_hidden_file
    settings = self.get_settings()
    searcher = Searcher.new(settings)
    f = '.gitignore'
    assert(!searcher.filter_file(f))
  end

  def test_filter_file_hidden_includehidden
    settings = self.get_settings()
    settings.excludehidden = false
    searcher = Searcher.new(settings)
    f = '.gitignore'
    assert(searcher.filter_file(f))
  end

  def test_filter_file_archive_no_searcharchives
    settings = self.get_settings()
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(!searcher.filter_file(f))
  end

  def test_filter_file_archive_searcharchives
    settings = self.get_settings()
    settings.searcharchives = 1
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(searcher.filter_file(f))
  end

  def test_filter_file_archive_archivesonly
    settings = self.get_settings()
    settings.archivesonly = true
    settings.searcharchives = true
    searcher = Searcher.new(settings)
    f = 'archive.zip'
    assert(searcher.filter_file(f))
  end

  def test_filter_file_nonarchive_archivesonly
    settings = self.get_settings()
    settings.archivesonly = true
    settings.searcharchives = true
    searcher = Searcher.new(settings)
    f = 'fileutil.rb'
    assert(!searcher.filter_file(f))
  end

################################################################################
# search_lines tests
################################################################################
  def test_search_lines
    settings = self.get_settings()
    searcher = Searcher.new(settings)
    testfile = self.get_test_file()
    contents = File.open(testfile, mode: 'r:ISO8859-1').read
    results = searcher.search_multiline_string(contents)
    assert_equal(results.count, 2)

    firstResult = results[0]
    assert_equal(firstResult.linenum, 23)
    assert_equal(firstResult.match_start_index, 3)
    assert_equal(firstResult.match_end_index, 11)

    secondResult = results[1]
    assert_equal(secondResult.linenum, 29)
    assert_equal(secondResult.match_start_index, 24)
    assert_equal(secondResult.match_end_index, 32)
  end

################################################################################
# search_multiline_string tests
################################################################################
  def test_search_multiline_string
    settings = self.get_settings()
    searcher = Searcher.new(settings)
    testfile = self.get_test_file()
    fo = File.open(testfile, mode: 'r:ISO8859-1')
    line_iterator = fo.each_line
    results = searcher.search_line_iterator(line_iterator)
    assert_equal(results.count, 2)

    firstResult = results[0]
    assert_equal(firstResult.linenum, 23)
    assert_equal(firstResult.match_start_index, 3)
    assert_equal(firstResult.match_end_index, 11)

    secondResult = results[1]
    assert_equal(secondResult.linenum, 29)
    assert_equal(secondResult.match_start_index, 24)
    assert_equal(secondResult.match_end_index, 32)
  end
end
